Add thanos query frontend sub command

yeya24 opened the pull request:
Signed-off-by: Ben Ye <yb532204897@gmail.com>

<!--
    Keep PR title verbose enough and add prefix telling
    about what components it touches e.g "query:" or ".*:"
-->

<!--
    Don't forget about CHANGELOG!

    Changelog entry format:
    - [#<PR-id>](<PR-URL>) Thanos <Component> ...

    <PR-id> Id of your pull request.
    <PR-URL> URL of your PR such as https://github.com/thanos-io/thanos/pull/<PR-id>
    <Component> Component affected by your changes such as Query, Store, Receive.
-->

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

<!-- Enumerate changes you made -->

The current code is almost the same as the Cortex frontend. I just removed the sharding middleware because seems that is specific for Cortex.

TODO:
- [ ] Implement Cortex cache.Cache interface, this should be done in another pr.
- [x] Unit tests and E2E tests.
- [x] Support parsing Thanos specific query parameters such as `dedup`, `max_source_resolution`, `partial_response`, etc. These params should be part of the cache key because these affect the query results. We need to implement Thanos codec to decode/encode Thanos query requests. https://github.com/yeya24/thanos/pull/159

## Verification

<!-- How you tested it? How do you know it works? -->

bwplotka commented on cmd/thanos/query-frontend.go:
```diff
@@ -0,0 +1,174 @@
+	cmd.Flag("query-range.split-interval", "Split queries by an interval and execute in parallel, 0 disables it.").
+		Default("24h").DurationVar(&c.splitInterval)
+
+	cmd.Flag("query-range.max-retries-per-request", "Maximum number of retries for a single request; beyond this, the downstream error is returned.").
```
So retries does not work for us API. (sad), so we need to rework this and for now maybe remove this.

yeya24 replied:
This is used in query range as well https://github.com/cortexproject/cortex/blob/master/pkg/querier/queryrange/roundtrip.go#L192, not only in the queue mode. Can you please check it?

bwplotka replied:
You are right, was wrong (: Cool!

bwplotka commented on cmd/thanos/query-frontend.go:
```diff
@@ -0,0 +1,174 @@
+	cmd.Flag("query-range.max-retries-per-request", "Maximum number of retries for a single request; beyond this, the downstream error is returned.").
+		Default("5").IntVar(&c.maxRetries)
+
+	cmd.Flag("query-range.disable-step-align", "Disable aligning incoming queries' start and end with their step.").
```
Let's remove this option and only expose this if someone will need. In practice you really want true. Plus Grafana does that by default

yeya24 commented:
Thanos Query API supports many custom parameters that Prometheus doesn't support, like `dedup`, `partial_response`, `max_source_resolution`, `storeMatch`, etc.

IMHO Thanos should implement its own Request type just like the [loki request](https://github.com/grafana/loki/blob/master/pkg/querier/queryrange/codec.go). Is it better for me to include it in this pr, or in another pr?

brancz commented:
> IMHO Thanos should implement its own Request type just like the loki request. Is it better for me to include it in this pr, or in another pr?

I haven't had the chance to review this PR in itself but I think since this work will be on-going I think I would prefer to merge a minimum state that works good enough :tm:, and then iterate on it. So to be explicit, I am for this, but in a follow up PR.

bwplotka commented on cmd/thanos/query-frontend.go:
```diff
@@ -0,0 +1,207 @@
+
+func (c *responseCacheConfig) registerFlag(cmd *kingpin.CmdClause) {
+	c.fifoCache.registerFlag(cmd)
+	cmd.Flag("query-range.response-cache-max-freshness", "Most recent allowed cacheable result, to prevent caching very recent results that might still be in flux.").
```
I believe this should be overall general config not fifo? :thinking:

yeya24 replied:
Yes, you are right. It is a general cache config so I named it `query-range.response-cache-max-freshness`.

bwplotka commented on cmd/thanos/query-frontend.go:
```diff
@@ -0,0 +1,207 @@
+
+// fifoCacheConfig defines configurations for Cortex fifo cache.
+type fifoCacheConfig struct {
+	maxSizeBytes units.Base2Bytes
```
BTW I think we should have those as cache client now and use our cache flags like in: https://github.com/thanos-io/thanos/blob/82cca562ce9e04e6ed6cbef41be71fc4b2d88207/cmd/thanos/store.go#L54

yeya24 replied:
Done.

bwplotka commented on cmd/thanos/query-frontend.go:
```diff
@@ -0,0 +1,207 @@
+	m[comp.String()] = func(g *run.Group, logger log.Logger, reg *prometheus.Registry, _ opentracing.Tracer, _ <-chan struct{}, _ bool) error {
+
+		return runQueryFrontend(
+			g,
```
no need for this syle - I think it can fit a line (:

bwplotka commented on cmd/thanos/query-frontend.go:
```diff
@@ -0,0 +1,207 @@
+		conf.queryRangeConfig.respCacheConfig.cacheMaxFreshness,
+	)
+
+	// TODO(yeya24): support other cache when available.
```
:+1:

yeya24 commented:
Some build errors got after rebasing master.

```
# github.com/cortexproject/cortex/pkg/querier/queryrange
../../../go/pkg/mod/github.com/cortexproject/cortex@v1.2.1-0.20200805064754-d8edc95e2c91/pkg/querier/queryrange/value.go:92:2: invalid case parser.ValueTypeVector in switch on promRes.Data.ResultType (mismatched types parser.ValueType and string)
../../../go/pkg/mod/github.com/cortexproject/cortex@v1.2.1-0.20200805064754-d8edc95e2c91/pkg/querier/queryrange/value.go:92:2: invalid case parser.ValueTypeMatrix in switch on promRes.Data.ResultType (mismatched types parser.ValueType and string)
!! command failed: build -o /home/yeya24/go/bin/thanos -ldflags -X github.com/prometheus/common/version.Version=0.14.0 -X github.com/prometheus/common/version.Revision=f3b29ab18ede95baad53cda564b490916efb5407 -X github.com/prometheus/common/version.Branch=add-thanos-query-frontend -X github.com/prometheus/common/version.BuildUser=yeya24@yeya24 -X github.com/prometheus/common/version.BuildDate=20200807-21:34:12  -extldflags '-static' -a -tags netgo github.com/thanos-io/thanos/cmd/thanos: exit status 2
make: *** [Makefile:114: build] Error 1
```

Seems the Prometheus dependency needs to be updated on the Cortex side as well,

bwplotka commented:
That might be true. Here comes cyclic deps (:

bwplotka commented:
Let's propose a PR on their side cc @pracucci

yeya24 commented:
> Let's propose a PR on their side cc @pracucci

I will prepare a pr.

yeya24 commented on pkg/api/queryfrontend/v1.go:
```diff
@@ -0,0 +1,50 @@
+
+	r.Get("/labels", instr("labels", handleFunc))
+	r.Post("/labels", instr("labels", handleFunc))
+}
```
Is it necessary to add `stores` and `rules` APIs here?

bwplotka replied:
well depends how we handle this - I think query frontend was doing pass through for all not defined endpoints, but we should test against it. If that's the case then why we define labels, series, values if not cached?

bwplotka replied:
Can we check it? (:

yeya24 commented:
I have updated my pr to support the response cache config file. PTAL when you have time. For the Cortex dependency error, I opened a pr https://github.com/cortexproject/cortex/pull/3000 already.

bwplotka commented on CHANGELOG.md:
```diff
@@ -35,6 +35,7 @@ We use *breaking* word for marking changes that are not backward compatible (rel
 - [#2865](https://github.com/thanos-io/thanos/pull/2865) ui: Migrate Thanos Ruler UI to React
 - [#2964](https://github.com/thanos-io/thanos/pull/2964) Query: Add time range parameters to label APIs. Add `start` and `end` fields to Store API `LabelNamesRequest` and `LabelValuesRequest`.
 - [#2996](https://github.com/thanos-io/thanos/pull/2996) Sidecar: Add `reloader_config_apply_errors_total` metric. Add new flags `--reloader.watch-interval`, and `--reloader.retry-interval`.
+- [#2973](https://github.com/thanos-io/thanos/pull/2973) Add Thanos Query Frontend component.
```
Docs, would be nice, maybe in next PR

bwplotka commented on cmd/thanos/query-frontend.go:
```diff
@@ -0,0 +1,184 @@
+	cmd.Flag("query-range.max-query-parallelism", "Maximum number of queries will be scheduled in parallel by the frontend.").
+		Default("14").IntVar(&c.maxQueryParallelism)
+
+	cmd.Flag("query-range.response-cache-max-freshness", "Most recent allowed cacheable result, to prevent caching very recent results that might still be in flux.").
```
We need better help flag description - in separate PR is ok

bwplotka replied:
essentially let's describe why this is needed

yeya24 replied:
Make sense, we can also help update the flag description in Cortex as well.

bwplotka commented on pkg/queryfrontend/cache/inmemory.go:
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package cache
```
`inmem`?

bwplotka replied:
ah it's the same package... I think it's fine for now

bwplotka commented on pkg/queryfrontend/roundtrip.go:
```diff
@@ -0,0 +1,125 @@
+				op = labelQueryRange
+			}
+
+			queriesCount.WithLabelValues(op).Inc()
```
this is weird. I think we put this this tripperware for `query`, `query range` but also` /label` `series` `label/.../values`?

yeya24 replied:
Just refactored this part. Since we only keep `query` and `query_range` endpoints, so we don't need to think about other endpoints anymore.

bwplotka commented on pkg/queryfrontend/roundtrip_test.go:
```diff
@@ -0,0 +1,222 @@
+// TestRoundTripRetryMiddleware tests the retry middleware.
+func TestRoundTripRetryMiddleware(t *testing.T) {
+	testRequest := &queryrange.PrometheusRequest{
+		Path:  "/api/v1/query_range",
```
What if path is totally different? not query range? Can we test it?

bwplotka commented on pkg/queryfrontend/roundtrip_test.go:
```diff
@@ -0,0 +1,222 @@
+		{
+			name: "disable split",
+			req: &queryrange.PrometheusRequest{
+				Path:  "/api/v1/query_range",
```
again, can we test different paths? just to see what to expect

bwplotka commented on test/e2e/query_frontend_test.go:
```diff
@@ -0,0 +1,193 @@
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(1), "cortex_cache_fetched_keys"))
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(0), "cortex_cache_hits"))
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(1), "querier_cache_added_new_total"))
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(1), "querier_cache_added_total"))
```
can we check if backend (querier) was touched? (: using some grpc_client conn metric on querier (: WDYT?

yeya24 replied:
Good idea! Let me update it.

yeya24 replied:
What about using `grpc_server_handled_total{method="xxx"}`?

yeya24 replied:
I think maybe checking `http_requests_total` on querier is also fine?

bwplotka replied:
yup :+1:

bwplotka commented:
Also still somehow tests flakes, wonder if it's SWIFT flakiness

yeya24 commented:
> `well depends how we handle this - I think query frontend was doing pass through for all not defined endpoints, but we should test against it. If that's the case then why we define labels, series, values if not cached?`?

Not passing through. Please see https://github.com/cortexproject/cortex/issues/2742. Now only defined endpoints are accessible. That's why the frontend needs to register other endpoints.

Only query range results are cached. The workflow is:
1. Non-query range but defined endpoints -> pass through to downstream URL. So no middlewares will be applied here.
2. query range -> go through all defined middlewares.

> `this is weird. I think we put this this tripperware for query, query range but also /label series label/.../values?`

Thanks, this metric is the same as Cortex. I agree we should separate labels for other endpoints as well.

yeya24 commented:
Seems we got another flaky test https://app.circleci.com/pipelines/github/thanos-io/thanos/2764/workflows/74964958-9ce1-4041-b3b9-4ea659a1764e/jobs/10997

bwplotka commented:
In this case let's do following:

* Use handler in query_range and query since we will soon modify it to use our tripperware
* For everything else let's have (SEPARATE!) passthrough handler. WDYT? (: Also I am surprised it does not work as our UI works just fine on query frontend :thinking: 
TBH I don't feel like we need allowlist all of endpoints. Maybe we can catch query + query range for now and pass through rest? 
Make it more transparent (:

yeya24 commented:
> In this case let's do following:
> 
> * Use handler in query_range and query since we will soon modify it to use our tripperware
> * For everything else let's have (SEPARATE!) passthrough handler. WDYT? (: Also I am surprised it does not work as our UI works just fine on query frontend thinking
>   TBH I don't feel like we need allowlist all of endpoints. Maybe we can catch query + query range for now and pass through rest?
>   Make it more transparent (:

Hello, I agree that we shouldn't limit the endpoints here and it is better to pass through all other routes to downstream querier to make sure everything works fine (like Grafana).

But I just found it not as easy as I thought to deal with the default route with `github.com/prometheus/common/route` package, so I removed the router and added the routing logic to the tripperware. WDYT? This way we don't need to have other routers. This works but TBH I don't know whether this is a good design or not.

``` go
	return func(next http.RoundTripper) http.RoundTripper {
		queryRangeTripper := queryrange.NewRoundTripper(next, codec, queryRangeMiddleware...)
		return frontend.RoundTripFunc(func(r *http.Request) (*http.Response, error) {
			switch r.URL.Path {
			case "/api/v1/query":
				if r.Method == http.MethodGet || r.Method == http.MethodPost {
					queriesCount.WithLabelValues(labelQuery).Inc()
				}
			case "api/v1/query_range":
				if r.Method == http.MethodGet || r.Method == http.MethodPost {
					queriesCount.WithLabelValues(labelQueryRange).Inc()
					return queryRangeTripper.RoundTrip(r)
				}
			default:
			}
			return next.RoundTrip(r)
		})
	}, nil
```

yeya24 commented:
PR is updated and I added more test cases. PTAL

bwplotka commented:
Happy with whatever makes sense more (: Will look today

bwplotka commented on cmd/thanos/query-frontend.go:
```diff
@@ -0,0 +1,187 @@
+			})
+			return hf
+		}
+		srv.Handle("/", injectf(fe.Handler().ServeHTTP))
```
:+1:

bwplotka commented on test/e2e/query_frontend_test.go:
```diff
@@ -0,0 +1,236 @@
+
+	t.Run("same range query, cache hit.", func(t *testing.T) {
+		// Run the same range query again, the result can be retrieved from cache directly.
+		rangeQuery(
```
can we check that other endpoints are pass through?

yeya24 replied:
I already tested `labelNames` and `labelValues` here https://github.com/thanos-io/thanos/blob/master/test/e2e/query_frontend_test.go#L99-L118. Is it not enough?
